### PR TITLE
In memory metrics: Keep track of created metrics [PLEN-98]

### PR DESCRIPTION
**Pull request by nicu-da**
This allows us in test to read all the created metrics and find metrics for which we don't have a reference (or that are created by third parties)

### CODE REVIEW

File: observability/metrics/src/test/lib/scala/com/daml/metrics/api/testing/InMemoryMetricsFactory.scala [hunk 1]
```diff
@@ -21,48 +21,106 @@ import com.daml.metrics.api.testing.InMemoryMetricsFactory.{
...
-  ): MetricHandle.Gauge[T] = InMemoryGauge(context, initial)
+  ): MetricHandle.Gauge[T] = metrics.addGauge(name, context, InMemoryGauge(context, initial))
 
   override def gaugeWithSupplier[T](
       name: MetricName,
       gaugeSupplier: () => T,
       description: String,
-  )(implicit context: MetricsContext): Unit = asyncGauges.addOne(name -> context -> gaugeSupplier)
+  )(implicit context: MetricsContext): Unit =
+    metrics.asyncGauges.addOne(name -> context -> gaugeSupplier)
```
Comment (lucb-da) [inline observability/metrics/src/test/lib/scala/com/daml/metrics/api/testing/InMemoryMetricsFactory.scala:57]: [No action required] I understand that it is not really needed, but it is a bit strange that gaugeWithSupplier doesn't follow the pattern of the other metrics: a `GaugeWithSupplier` class, and a `metrics.addGaugeWithSupplier` method.

File: observability/metrics/src/test/lib/scala/com/daml/metrics/api/testing/InMemoryMetricsFactory.scala [hunk 2]
```diff
@@ -21,48 +21,106 @@ import com.daml.metrics.api.testing.InMemoryMetricsFactory.{
...
 
-  val asyncGauges: ConcurrentMap[(MetricName, MetricsContext), () => Any] =
-    TrieMap[(MetricName, MetricsContext), () => Any]()
+  val metrics: MetricsState =
+    MetricsState(
+      timers = mutable.Map.empty,
+      gauges = mutable.Map.empty,
+      meters = mutable.Map.empty,
+      counters = mutable.Map.empty,
+      histograms = mutable.Map.empty,
```
Comment (lucb-da) [inline observability/metrics/src/test/lib/scala/com/daml/metrics/api/testing/InMemoryMetricsFactory.scala:40]: I don't think we have any guaranty that multiple metrics won't be added at the same time. So those maps likely should be `TrieMap` too.

File: observability/metrics/src/test/lib/scala/com/daml/metrics/api/testing/InMemoryMetricsFactory.scala [hunk 3]
```diff
@@ -21,48 +21,106 @@ import com.daml.metrics.api.testing.InMemoryMetricsFactory.{
...
 object InMemoryMetricsFactory extends InMemoryMetricsFactory {
 
-  private def addToContext[T](
+  type MetricIdentifier = (MetricName, MetricsContext)
+  case class MetricsState(
+      timers: mutable.Map[MetricIdentifier, InMemoryTimer],
+      gauges: mutable.Map[MetricIdentifier, InMemoryGauge[_]],
+      meters: mutable.Map[MetricIdentifier, InMemoryMeter],
+      counters: mutable.Map[MetricIdentifier, InMemoryCounter],
+      histograms: mutable.Map[MetricIdentifier, InMemoryHistogram],
```
Comment (lucb-da) [inline observability/metrics/src/test/lib/scala/com/daml/metrics/api/testing/InMemoryMetricsFactory.scala:81]: [no action required] Having the metric contexts split between the maps in here, and the maps in the individual metrics feel a bit incomplete to me.
I can see a point where we only have maps in `MetricsState`. But it also might make looking for values with a specific set of labels more complexe ...
I guess we will adapte it when we need it.

**Review by lucb-da — COMMENTED** [review]
It looks fine.

File: observability/metrics/src/test/lib/scala/com/daml/metrics/api/testing/InMemoryMetricsFactory.scala (hunk 2 above)
Comment (nicu-da) [reply observability/metrics/src/test/lib/scala/com/daml/metrics/api/testing/InMemoryMetricsFactory.scala]: Good point, migrated everything to concurrent maps

**Review by lucb-da — APPROVED** [review]
Let's go.

**Pull request merged**